Bottom: 356f33f3c593f1a18af5fa0b9c4028b8124d03fb
Top:    b6ec0d911e018e1904617006d1f312fc46dd91a8
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-09-05 15:17:49 +0800

Refresh of vhost-avoid-unnecessary-iov

---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index 9e8d1f3..d6d27c6 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -527,16 +527,18 @@ static void handle_tx(struct vhost_net *net)
 
 			/* Skip header. TODO: support TSO. */
 			len = iov_length(vq->iov, out);
-			iov_iter_init(&msg.msg_iter, WRITE, vq->iov, out, len);
-			iov_iter_advance(&msg.msg_iter, hdr_size);
-			/* Sanity check */
-			if (!msg_data_left(&msg)) {
-				vq_err(vq, "Unexpected header len for TX: "
-					"%zd expected %zd\n",
-					len, hdr_size);
-				goto out;
+			iov_iter_init(&msg.msg_iter, WRITE,  vq->iov, out, len);
+			if (unlikely(hdr_size)) {
+				iov_iter_advance(&msg.msg_iter, hdr_size);
+				/* Sanity check */
+				if (!msg_data_left(&msg)) {
+					vq_err(vq, "Unexpected header len for "
+						   "TX: %zd expected %zd\n",
+						len, hdr_size);
+					goto out;
+				}
+				len = msg_data_left(&msg);
 			}
-			len = msg_data_left(&msg);
 
 			zcopy_used = zcopy && len >= VHOST_GOODCOPY_LEN
 				     && (nvq->upend_idx + 1) % UIO_MAXIOV !=
